File: agclib/PCM.cpp
#include <cmath>

#include "PCM.h"

namespace agc {
namespace pcm {

double calculatePeakAmplitude(const std::vector<char>& pcm)
{
	double peak = 0;
	for (size_t i = 0; i < pcm.size(); i += sizeof(short)) {
		short sample = std::abs(*reinterpret_cast<const short*>(&pcm[i]));
		if (peak < sample) {
			peak = sample;
		}
	}

	// Normalize to 0-1 scale
	return peak / (1 << 15);
}

}
}

File: agclib/src/pcm.rs
/// Maximum absolute value of a 16-bit signed PCM sample.
const MAX_AMPLITUDE: f64 = (1u32 << 15) as f64;

/// Returns the peak amplitude of a 16-bit little-endian PCM buffer,
/// normalised to `[0.0, 1.0]`. Returns `0.0` for an empty buffer.
pub fn calculate_peak_amplitude(pcm: &[u8]) -> f64 {
    let peak = pcm
        .chunks_exact(size_of::<i16>())
        .map(|bytes| {
            // SAFETY: chunks_exact guarantees exactly size_of::<i16>() bytes.
            i16::from_ne_bytes(bytes.try_into().unwrap()).unsigned_abs()
        })
        .max()
        .unwrap_or(0);

    peak as f64 / MAX_AMPLITUDE
}
